File: tic-tac-toe/algorithm.py
from itertools import groupby
from typing import Iterable

def init_board(n: int) -> list[list[str]]:
  return [['' for _ in range(n)] for _ in range(n)]

def check(arr: Iterable[Iterable[str]]) -> bool:
  def check_row(r: Iterable[str]) -> bool:
    g = groupby(r)
    # Return True if row' or column' elements match eacother, return False if empty spot is present
    return next(g, True) and not next(g, False) if '' not in r else False
  
  range_ = range(len(arr))

  # Check rows and colums
  for i in range_:
    column = [j[i] for j in arr]
    if check_row(column) or check_row(arr[i]): return True
  
  # Check diagonals
  first_diagonal  = [arr[i][i]      for i in range_] # ↘️
  second_diagonal = [arr[i][-1 - i] for i in range_] # ↙️
  if check_row(first_diagonal) or check_row(second_diagonal): return True
  
  return False

board = init_board(3)

File: tic-tac-toe/algorithm.ts


const init_board = (n: number): string[][] => {
  const board: string[][] = []
  
  for (let i = 0; i < n; i++) {
    board.push([])
    for (let j = 0; j < n; j++) board[i].push('')
  }
  
  return board
}

const check = (arr: string[][]): boolean => {
  const range: number = arr.length
  const check_arr = (arr: string[]): boolean => {
    return arr.every(e => e === arr[0]) && !arr.includes('')
  }

  for (let i = 0; i < range; i++) {
    const column: string[] = []
    for (let j = 0; j < range; j++) column.push(arr[j][i])
    if (check_arr(column) || check_arr(arr[i])) return true
  }

  const first_diag: string[] = []
  const second_diag: string[] = []
  for (let i = 0; i < range; i++) {
    first_diag.push(arr[i][i])
    second_diag.push(arr[i][range - 1 - i])
  }
  if (check_arr(first_diag) || check_arr(second_diag)) return true
  
  return false
}

const board = init_board(3)
